Transform propagator and compute at map out of lowering.

csarofeen opened the pull request:
Make a traversal pass to transform the fusion consistent with a tensor.

csarofeen committed:
Final compute at cleanup.

csarofeen committed:
Add compute at modes BestEffort and MostInlined.

csarofeen committed:
Fix expression sorting add a test that was broken.

csarofeen committed:
Clang format.

csarofeen committed:
Minor fix to expr ordering. Will be needed for most inlined on normalization.

csarofeen committed:
Validate parallel type for split and merge on TensorView interface.

csarofeen committed:
Merge branch 'split_merge_fix' into compute_at_modes

csarofeen committed:
Move normalization to most inlined.

csarofeen committed:
Draft transform propagator in a more usable manner. Place it with TransformReplay.

csarofeen committed:
Move lower_compute_at_map to compute_at_map.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower2device.cpp:
```diff
@@ -356,7 +356,6 @@ kir::Expr* GpuLower::lowerExpr(const Expr* expr) {
 }
 
 GpuLower* GpuLower::current() {
-  TORCH_INTERNAL_ASSERT(active_gpu_lower != nullptr);
```
Is this reasonable, or should we keep this assert and send an argument to the compute at map whether or not its a lowering map?

csarofeen commented:
Note: https://github.com/csarofeen/pytorch/pull/676/commits/95a03712b0d4f28c76dac333b78a0fe100005d00

naoyam commented on torch/csrc/jit/codegen/cuda/compute_at_map.h:
```diff
@@ -71,6 +71,8 @@ class TORCH_CUDA_CU_API ComputeAtMap {
   std::string toString();
 
  private:
+  bool has_lowered_kir_ = false;
```
I assume the intention is to make `ComputeAtMap` available without lowering. Wouldn't it be simpler and cleaner if the mappings for KIR nodes are completely separated from the mappings for the Fusion IR? For example, how about letting `ComputeAtMap` only take care of the Fusion IR nodes, and create, say, class `kir::ComputeAtMap`, that is "lowered" from `ComputeAtMap`?

naoyam replied:
@csarofeen  I updated the interface just a little bit in order to make things more explicit. `build` takes `Fusion*` and `GpuLower*` as its parameters. If `GpuLower*` is `nullptr`, it doesn't generate the KIR version of the mappings.

https://github.com/csarofeen/pytorch/pull/688/commits/94ee1ecf9664ed44f764074a141553f34bca80bf

Commit:
suppress clang-tidy warnings

Commit:
Merge branch '20_12_3_devel' into transform_propagator

Commit:
Simple refactoring of ComputeAtMap

Pass Fusion and GpuLower pointers explicitly instead of using the
"current" pointers. The intention is to make the behavior explicit.
